booster/library/Booster/Pattern/Binary.hs

```diff
@@ -401,37 +513,58 @@ decodeMagicHeaderAndVersion = do
             pure Nothing
     hdrSize = 19
 
-supported :: Version -> Bool
-supported version = version.major == 1 && version.minor `elem` [0 .. 2]
+    supported :: Version -> Bool
+    supported version = version.major == 1 && version.minor `elem` [0 .. 2]
 
 decodeTerm' :: Maybe KoreDefinition -> Get Term
 decodeTerm' mDef = do
     (version, mbSize) <- decodeMagicHeaderAndVersion
-    runDecodeM version mDef (decodeBlock mbSize) >>= \case
-        [BTerm trm] -> pure trm
-        _ -> fail "Expecting a single term on the top of the stack"
+    runDecodeM version mDef $
+        decodeBlock mbSize >>= \case
+            [BTerm trmIdx] -> fmap stripRawTerm $ getTerm trmIdx
```

Comment:
Seems like magic to me!

We have a single call to `decodeTerm` in the `Booster.LLVM` module, as far as I know. Do I understand correctly that the cache lifetime is confined to `decodeTerm'`?

Also, the `getTerm trmIdx` call will return the fully resolved term? The subterms of that term may be present in the store before, but only once?

I think I know the answers, just want a confirmation.

Big brain @jberthold !

Reply:
> We have a single call to `decodeTerm` in the `Booster.LLVM` module, as far as I know. Do I understand correctly that the cache lifetime is confined to `decodeTerm'`?

Correct, the cache starts empty in `runDecodeM` called here. We could carry a cache around across calls but then we would have to check for any unpacked term whether or not we already have it in the cache (term comparisons, but not recursive deep terms). This is good enough I think.

> Also, the `getTerm trmIdx` call will return the fully resolved term? The subterms of that term may be present in the store before, but only once?

Yes, the term is already fully resolved in `registerTerm` before it is stored in the `termStore` (...hopefully, because otherwise any inconsistent data would sit there as a thunk waiting to explode).
 
> I think I know the answers, just want a confirmation.
> 
> Big brain @jberthold !

:blush: not my idea originally, the credit goes to Virgil.